### PR TITLE
Fix a bug with inferior calls by tracking volatile mode state.

### PR DESCRIPTION
**Pull request by mark-undoio**
A minor (obvious in hindsight) fix.  This is covered by the Delve test suite, I don't have an internal end-to-end test for it yet.

### CODE REVIEW

File: pkg/proc/gdbserial/undo.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ import (
 type undoSession struct {
 	checkpointNextId int
 	checkpoints      map[int]proc.Checkpoint
+	volatile         bool
```
Comment (gareth-rees) [inline pkg/proc/gdbserial/undo.go]: Could you add a comment documenting the new field?

File: pkg/proc/gdbserial/undo.go [hunk 2]
```diff
@@ -324,6 +326,21 @@ func (uc *undoSession) travelToTime(p *gdbProcess, pos string) error {
 	return err
 }
 
+func (uc *undoSession) activateVolatile(p *gdbProcess) (func(), error) {
```
Comment (gareth-rees) [inline pkg/proc/gdbserial/undo.go:341]: Could you add a comment documenting the new method?

**Review by gareth-rees — COMMENTED** [review]
This looks ok, though it's hard for me to review because I don't know how volatile mode works in Delve. For example, you've made it an error to restart in volatile mode. I can't tell if that represents a user error (the user should not have tried to restart in this mode) or a Delve programming error (Delve should have exited volatile mode before calling `Restart`). If these cases represent Delve programming errors then maybe assertions would be more appropriate than error propagation, or if that's not how Delve works, then maybe phrase the error message to make clear that some internal invariant has been violated.

File: pkg/proc/gdbserial/undo.go (hunk 1 above)
Comment (mark-undoio) [reply pkg/proc/gdbserial/undo.go]: :+1:

File: pkg/proc/gdbserial/undo.go (hunk 2 above)
Comment (mark-undoio) [reply pkg/proc/gdbserial/undo.go:341]: :+1:

**Comment by mark-undoio** [comment]
> This looks ok, though it's hard for me to review because I don't know how volatile mode works in Delve. For example, you've made it an error to restart in volatile mode. I can't tell if that represents a user error (the user should not have tried to restart in this mode) or a Delve programming error (Delve should have exited volatile mode before calling `Restart`). If these cases represent Delve programming errors then maybe assertions would be more appropriate than error propagation, or if that's not how Delve works, then maybe phrase the error message to make clear that some internal invariant has been violated.

I think that would be a Delve programming error.  I've changed these cases (along with a few on the other PR you reviewed) to use `panic` instead of returning a normal error.

The semantics of `panic` in Go are different to our C/C++ product's idea of panic, since they are recoverable.  By putting an unexpected panic in here I think it gives a source-level hint that something is wrong, plus more appropriate user-visible behaviour.

It is possible to catch a panic and handle it at a higher layer, which I think Delve will actually do (so you'll get a message from the CLI about it, rather than a complete explosion).

**Comment by mark-undoio** [comment]
This was released as part of `undo-1.20.2-4`

**Pull request closed without merging**